java-frontend/src/test/java/org/sonar/java/model/JParserSemanticTest.java

```diff
@@ -349,6 +349,17 @@ void declaration_type() {
     assertThat(cu.sema.declarations.get(c.typeBinding)).isSameAs(c);
   }
 
+  @Test
+  void declaration_type_parameterized() {
+    JavaTree.CompilationUnitTreeImpl cu = test("class C<T> { C<String> field; }");
+    ClassTreeImpl c = (ClassTreeImpl) cu.types().get(0);
+    AbstractTypedTree fieldType = (AbstractTypedTree) ((VariableTreeImpl) c.members().get(0)).type();
+
+    assertThat(c.typeBinding).isNotNull();
+    assertThat(cu.sema.declarations.get(c.typeBinding)).isSameAs(c);
+    assertThat(cu.sema.usages.get(c.typeBinding)).containsOnly((IdentifierTreeImpl) ((ParameterizedTypeTree) fieldType).type());
```

Comment:
Modification in `JParser#usage` is enough to pass this test, so I'm wondering if other modifications are really needed.

To use modifications in `JSymbol` one need to call `declaration` or `usages` on `cu.sema.typeSymbol(fieldType.typeBinding)`, which corresponds to `fieldType.symbolType().symbol()` in old implementation. And I'm wondering about existence of such calls.

Reply:
I can see quite a lot of `symbolType().symbol()` in checks (14 matches, 13 checks relying on this).

Reply:
However modifications in `JSymbol` are mandatory, otherwise we simply never get back to the declaration bindings from parameterized symbol. I didn't tested it because calling `symbol().usages()` in `JParserSemanticTest` returns the old symbol.

Reply:
Call `cu.sema.typeSymbol(fieldType.typeBinding).usages()` :wink: